File: android/model/services/DatabaseService.kt
package com.tobiasheymann.ep.model.services

import com.tobiasheymann.ep.model.News
import com.tobiasheymann.ep.model.Plant
import de.thm.tp.library.firebase.firestore.TPFirebaseFirestore
import de.thm.tp.library.firebase.firestore.TPFirebaseFirestoreQueryBuilder

class DatabaseService {
    companion object {
        fun getLastTenNews(callback: (List<News>?) -> Unit) {
            TPFirebaseFirestore.addCollectionSnapshotListener(
                TPFirebaseFirestoreQueryBuilder(News.COLLECTION_NAME)
                    .orderBy("metadata.timestamp", true)
                    .limit(10)
            ) { result, error ->
                result?.let { resultObject ->
                    val list = mutableListOf<News>()
                    resultObject.forEach { tpFirebaseFirestoreDocument ->
                        News.toObject(tpFirebaseFirestoreDocument.data)?.let { news ->
                            list.add(news)
                        }
                    }
                    callback(list)
                }
                if (error != null) {
                    callback(null)
                }
            }
        }

        fun getLastFivePlants(callback: (List<Plant>?) -> Unit) {
            TPFirebaseFirestore.addCollectionSnapshotListener(
                TPFirebaseFirestoreQueryBuilder(Plant.COLLECTION_NAME)
                    .orderBy("metadata.timestamp", true)
                    .limit(5)
            ) { result, error ->
                result?.let { resultObject ->
                    val list = mutableListOf<Plant>()
                    resultObject.forEach { tpFirebaseFirestoreDocument ->
                        Plant.toObject(tpFirebaseFirestoreDocument.data)?.let { plant ->
                            list.add(plant)
                        }
                    }
                    callback(list)
                }
                if (error != null) {
                    callback(null)
                }
            }
        }

        fun getAllPlants(callback: (List<Plant>) -> Unit) {
            TPFirebaseFirestore.getDocuments(Plant.COLLECTION_NAME) { result, error ->
                result?.let { resultObject ->
                    val list = mutableListOf<Plant>()
                    resultObject.forEach { tpFirebaseFirestoreDocument ->
                        Plant.toObject(tpFirebaseFirestoreDocument.data)?.let { plant ->
                            list.add(plant)
                        }
                    }
                    callback(list)
                }
                if (error != null) {
                    callback(listOf())
                }
            }
        }

        fun searchPlantById(uuid: String, callback: (Plant) -> Unit) {
            TPFirebaseFirestore.getDocument("plants", uuid) { result, error ->
                result?.let { resultObject ->
                    Plant.toObject(resultObject.data)?.let { callback(it) }
                }
            }
        }
    }
}

File: ios/model/Services/DatabaseService.swift
class DatabaseService {
    static func getLastTenNews(callback: @escaping ([News]?) -> ()) {
        TPFirebaseFirestore.addCollectionSnapshotListener(queryBuilder: TPFirebaseFirestoreQueryBuilder(collectionName: News.COLLECTION_NAME).orderBy(field: "metadata.timestamp", descending: true).limit(limit: 10)) {
            result, error in
            if let resultObject = result {
                var list = [News]()
                resultObject.forEach {
                    tpFirebaseFirestoreDocument in
                    if let news = News.toObject(map: tpFirebaseFirestoreDocument.data) {
                        list.append(news)
                    }
                }

                callback(list)
            }

            if error != nil {
                callback(nil)
            }
        }
    }

    static func getLastFivePlants(callback: @escaping ([Plant]?) -> ()) {
        TPFirebaseFirestore.addCollectionSnapshotListener(queryBuilder: TPFirebaseFirestoreQueryBuilder(collectionName: Plant.COLLECTION_NAME).orderBy(field: "metadata.timestamp", descending: true).limit(limit: 5)) {
            result, error in
            if let resultObject = result {
                var list = [Plant]()
                resultObject.forEach {
                    tpFirebaseFirestoreDocument in
                    if let plant = Plant.toObject(map: tpFirebaseFirestoreDocument.data) {
                        list.append(plant)
                    }
                }

                callback(list)
            }

            if error != nil {
                callback(nil)
            }
        }
    }

    static func getAllPlants(callback: @escaping ([Plant]) -> ()) {
        TPFirebaseFirestore.getDocuments(collectionName: Plant.COLLECTION_NAME) {
            result, error in
            if let resultObject = result {
                var list = [Plant]()
                resultObject.forEach {
                    tpFirebaseFirestoreDocument in
                    if let plant = Plant.toObject(map: tpFirebaseFirestoreDocument.data) {
                        list.append(plant)
                    }
                }

                callback(list)
            }

            if error != nil {
                callback([])
            }
        }
    }

    static func searchPlantById(uuid: String, callback: @escaping (Plant) -> ()) {
        TPFirebaseFirestore.getDocument(collectionName: "plants", documentId: uuid) {
            result, error in
            if let resultObject = result {
                if let it = Plant.toObject(map: resultObject.data) {
                    callback(it)
                }
            }
        }
    }
}
